core/tauri/src/ipc/protocol.rs

```diff
@@ -0,0 +1,272 @@
+// Copyright 2019-2023 Tauri Programme within The Commons Conservancy
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-License-Identifier: MIT
+
+use http::{
+  header::{ACCESS_CONTROL_ALLOW_HEADERS, ACCESS_CONTROL_ALLOW_ORIGIN},
+  HeaderValue, Method, StatusCode,
+};
+
+use crate::{
+  manager::WindowManager,
+  runtime::http::{Request as HttpRequest, Response as HttpResponse},
+  window::{InvokeRequest, UriSchemeProtocolHandler},
+  Runtime,
+};
+
+use super::{CallbackFn, InvokeBody, InvokeResponse};
+
+#[cfg(not(ipc_custom_protocol))]
+pub fn message_handler<R: Runtime>(
+  manager: WindowManager<R>,
+) -> crate::runtime::webview::WebviewIpcHandler<crate::EventLoopMessage, R> {
+  Box::new(move |window, request| handle_ipc_message(request, &manager, &window.label))
+}
+
+pub fn get<R: Runtime>(manager: WindowManager<R>, label: String) -> UriSchemeProtocolHandler {
+  Box::new(move |request| {
+    let mut response = match *request.method() {
+      Method::POST => {
+        let (mut response, content_type) = match handle_ipc_request(request, &manager, &label) {
+          Ok(data) => match data {
+            InvokeResponse::Ok(InvokeBody::Json(v)) => (
+              HttpResponse::new(serde_json::to_vec(&v)?.into()),
+              "application/json",
+            ),
+            InvokeResponse::Ok(InvokeBody::Raw(v)) => {
+              (HttpResponse::new(v.into()), "application/octet-stream")
+            }
+            InvokeResponse::Err(e) => {
+              let mut response = HttpResponse::new(serde_json::to_vec(&e.0)?.into());
+              response.set_status(StatusCode::BAD_REQUEST);
+              (response, "text/plain")
+            }
+          },
+          Err(e) => {
+            let mut response = HttpResponse::new(e.as_bytes().to_vec().into());
+            response.set_status(StatusCode::BAD_REQUEST);
+            (response, "text/plain")
+          }
+        };
+
+        response.set_mimetype(Some(content_type.into()));
+
+        response
+      }
+
+      Method::OPTIONS => {
+        let mut r = HttpResponse::new(Vec::new().into());
+        r.headers_mut().insert(
+          ACCESS_CONTROL_ALLOW_HEADERS,
+          HeaderValue::from_static("Content-Type, Tauri-Callback, Tauri-Error, Tauri-Channel-Id"),
```

Comment:
@tweidinger @chippers can I set the allow headers to "*" here? The headers can be accessed like this:

```rust
#[command]
fn my_cmd(request: tauri::ipc::Request) {
  let headers = request.headers();
}
```

Reply:
IMHO if there are no secrets meant to be sent via headers it is totally fine as default.
In the default case there are only the `Content-Type`, `Tauri-Callback`, `Tauri-Error`, `Tauri-Channel-Id` header set. These are not containing secret values from my perspective. Leaking the `Tauri-Channel-Id` could be abused but I can not really think of a scenario where it could be leaked without already having more privileges.

If there are going to be secrets in the headers or cookies attached to the request the developer would need to modify the `ACCESS_CONTROL_ALLOW_HEADERS` to suit their usage as the `*` value will be treated as a header name instead of a wildcard.

Can the developer modify this value at runtime or via configuration?
It doesn't seem supported on webkit https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers#browser_compatibility

Reply:
we can add an API in Rust for this @tillmann-crabnebula though I'd wait for it to be requested actually.